src/test/kotlin/InputValidationTest.kt

```
@@ -0,0 +1,53 @@
+import exception.NameException
+import exception.NullException
+import exception.TryCountException
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+class InputValidationTest {
+    private val nullException = NullException()
+    private val nameException = NameException()
+    private val tryCountException = TryCountException()
```

Exception이 아닌데 Exception처럼 보여서 오해를 일으키는 네이밍입니다.

---

다음은 코틀린에서의 예외처리 방식입니다.
당장은 바로 적응하기 힘들겠지만 장기적으로는 아래의 방식을 고려해주세요.

- [논리 오류](https://ko.wikipedia.org/wiki/%EB%85%BC%EB%A6%AC_%EC%98%A4%EB%A5%98)일 때만 예외를 던지세요.
- 논리 오류가 아니면 예외를 던지지 말고 null을 반환하세요.
- 실패하는 경우가 복잡해서 null로 처리할 수 없으면 sealed class를 반환하세요.
- 일반적인 코틀린 코드에서 `try-catch`를 사용하지 마세요.

위 사항들의 자세한 설명은 아래 글을 읽어보세요.
https://medium.com/@elizarov/kotlin-and-exceptions-8062f589d07

---

논리 오류가 의미하는 것이 프로그램 실행은 되지만 논리적으로 오류가 있는 것을 의미하는 것으로 알고있는데, 제작자가 설정한 프로그램 규칙에 위배되는 경우도  논리 오류라고 할 수 있나요? 예를 들면 자동차이름이 5글자가 넘어가는 경우, 에러를 던져야 하는지 아니면 null을 반환해야 하는지 잘 모르겠습니다. 그리고 입출력 상황에서 예외가 발생하였을 때, 재입력을 받고 싶은 경우, try catch 문을 사용해야만 할 거 같은데 다른 방법이 있을까요?

---

> 논리 오류가 의미하는 것이 프로그램 실행은 되지만 논리적으로 오류가 있는 것을 의미하는 것으로 알고있는데, 제작자가 설정한 프로그램 규칙에 위배되는 경우도 논리 오류라고 할 수 있나요? 

네, 논리 오류라고 볼 수 있습니다.
다만 논리 오류일 때, 예외를 던지라는 것은 fail fast를 위한 장치일 뿐이지 이미 그 전에 걸러져야 합니다.

> 예를 들면 자동차이름이 5글자가 넘어가는 경우, 에러를 던져야 하는지 아니면 null을 반환해야 하는지 잘 모르겠습니다. 

`자동차 이름이 5글자가 넘어가는 경우는 재입력을 받는다.`와 같은 요구사항에서 5글자가 넘는 경우는 충분히 예상 가능한 상황이기 때문에 예외 상황이라고 볼 수 없습니다.
따라서 null을 반환하거나 if문을 통해 다시 입력을 받도록 처리해야 합니다. 예외는 정말 예외적인 상황에서만 던져야 합니다.

> 그리고 입출력 상황에서 예외가 발생하였을 때, 재입력을 받고 싶은 경우, try catch 문을 사용해야만 할 거 같은데 다른 방법이 있을까요?

일반적인 코틀린 코드에서 `try-catch`를 사용하지 말라는 뜻은 도메인 로직이나 충분히 예상 가능한 상황들의 Flow를 처리할 때 사용하지 말라는 뜻입니다.
입출력 상황에서 발생하는 예외는 우리가 어떻게 할 수 있는 부분이 아니기 때문에 `try-catch`로 작성하셔도 상관없습니다.

대신 이러한 부분들도 null로 변경해서 코딩하시려면 `runCatching { ... }.getOrNull()`을 사용하시면 됩니다.